File: server/src/routes/codeChat.ts
import { Router } from 'express';
import { codeChatService } from '../services/codeChat.js';

const router = Router();

// Get available code-related knowledge bases
router.get('/knowledge-bases', async (req, res) => {
  try {
    const knowledgeBases = await codeChatService.getCodeKnowledgeBases();
    res.json(knowledgeBases);
  } catch (error) {
    console.error('Error fetching code knowledge bases:', error);
    res.status(500).json({ error: 'Failed to fetch knowledge bases' });
  }
});

// List all chat sessions
router.get('/sessions', async (req, res) => {
  try {
    const userId = req.query.userId ? parseInt(req.query.userId as string) : undefined;
    const sessions = await codeChatService.listSessions(userId);
    res.json(sessions);
  } catch (error) {
    console.error('Error listing chat sessions:', error);
    res.status(500).json({ error: 'Failed to list chat sessions' });
  }
});

// Create a new chat session
router.post('/sessions', async (req, res) => {
  try {
    const { userId, knowledgeBaseIds, title } = req.body;

    if (!knowledgeBaseIds || !Array.isArray(knowledgeBaseIds) || knowledgeBaseIds.length === 0) {
      return res.status(400).json({ error: 'At least one knowledge base must be selected' });
    }

    const session = await codeChatService.createSession({
      userId,
      knowledgeBaseIds,
      title,
    });

    res.status(201).json(session);
  } catch (error) {
    console.error('Error creating chat session:', error);
    const message = error instanceof Error ? error.message : 'Failed to create chat session';
    res.status(500).json({ error: message });
  }
});

// Get a specific chat session with messages
router.get('/sessions/:id', async (req, res) => {
  try {
    const id = parseInt(req.params.id);
    const result = await codeChatService.getSession(id);

    if (!result) {
      return res.status(404).json({ error: 'Chat session not found' });
    }

    res.json(result);
  } catch (error) {
    console.error('Error getting chat session:', error);
    res.status(500).json({ error: 'Failed to get chat session' });
  }
});

// Delete a chat session
router.delete('/sessions/:id', async (req, res) => {
  try {
    const id = parseInt(req.params.id);
    const deleted = await codeChatService.deleteSession(id);

    if (!deleted) {
      return res.status(404).json({ error: 'Chat session not found' });
    }

    res.status(204).send();
  } catch (error) {
    console.error('Error deleting chat session:', error);
    res.status(500).json({ error: 'Failed to delete chat session' });
  }
});

// Update session knowledge bases
router.patch('/sessions/:id/knowledge-bases', async (req, res) => {
  try {
    const id = parseInt(req.params.id);
    const { knowledgeBaseIds } = req.body;

    if (!knowledgeBaseIds || !Array.isArray(knowledgeBaseIds) || knowledgeBaseIds.length === 0) {
      return res.status(400).json({ error: 'At least one knowledge base must be selected' });
    }

    const session = await codeChatService.updateSessionKnowledgeBases(id, knowledgeBaseIds);

    if (!session) {
      return res.status(404).json({ error: 'Chat session not found' });
    }

    res.json(session);
  } catch (error) {
    console.error('Error updating session knowledge bases:', error);
    const message = error instanceof Error ? error.message : 'Failed to update session';
    res.status(500).json({ error: message });
  }
});

// Send a message to the chat session
router.post('/sessions/:id/messages', async (req, res) => {
  try {
    const id = parseInt(req.params.id);
    const { content } = req.body;

    if (!content || typeof content !== 'string' || content.trim().length === 0) {
      return res.status(400).json({ error: 'Message content is required' });
    }

    const result = await codeChatService.sendMessage(id, content.trim());

    res.status(201).json(result);
  } catch (error) {
    console.error('Error sending message:', error);
    const message = error instanceof Error ? error.message : 'Failed to send message';
    res.status(500).json({ error: message });
  }
});

export default router;

File: server/app/api/api_v1/endpoints/code_chat.py
from typing import List, Any, Optional, Dict
from fastapi import APIRouter, Depends, HTTPException, Response
from sqlmodel import Session
from app.core.db import get_session
from app.api.deps import get_current_user
from app.models.user import User
from app.models.code_chat import CodeChatSession, CodeChatMessage
from app.models.knowledge_base import KnowledgeBase
from app.services.code_chat_service import code_chat_service

router = APIRouter()

@router.get("/knowledge-bases", response_model=List[KnowledgeBase])
def list_knowledge_bases(
    session: Session = Depends(get_session),
    current_user: User = Depends(get_current_user)
) -> Any:
    return code_chat_service.get_code_knowledge_bases(session)

@router.get("/sessions", response_model=List[CodeChatSession])
def list_sessions(
    session: Session = Depends(get_session),
    current_user: User = Depends(get_current_user)
) -> Any:
    return code_chat_service.list_sessions(session, user_id=current_user.id)

@router.post("/sessions", response_model=CodeChatSession)
def create_session(
    data: Dict[str, Any],
    session: Session = Depends(get_session),
    current_user: User = Depends(get_current_user)
) -> Any:
    try:
        return code_chat_service.create_session(session, data, user_id=current_user.id)
    except ValueError as e:
        raise HTTPException(status_code=400, detail=str(e))

# ... (omitted lines)

@router.delete("/sessions/{id}")
def delete_session(
    id: int,
    session: Session = Depends(get_session),
    current_user: User = Depends(get_current_user)
) -> Any:
    if not code_chat_service.delete_session(session, id, user_id=current_user.id):
        raise HTTPException(status_code=404, detail="Session not found")
    return Response(status_code=204)

@router.post("/sessions/{id}/messages", response_model=Dict[str, CodeChatMessage])
def send_message(
    id: int,
    data: Dict[str, str],
    session: Session = Depends(get_session),
    current_user: User = Depends(get_current_user)
) -> Any:
    message = data.get("content") or data.get("message")
    if not message:
        raise HTTPException(status_code=400, detail="Message content is required")

    try:
        return code_chat_service.send_message(session, id, message, user_id=current_user.id)
    except ValueError as e:
        raise HTTPException(status_code=404, detail=str(e))
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))

@router.put("/sessions/{id}/knowledge-bases", response_model=CodeChatSession)
def update_session_knowledge_bases(
    id: int,
    data: Dict[str, List[int]],
    session: Session = Depends(get_session),
    current_user: User = Depends(get_current_user)
) -> Any:
    kb_ids = data.get("knowledgeBaseIds")
    if not kb_ids:
        raise HTTPException(status_code=400, detail="knowledgeBaseIds is required")

    updated = code_chat_service.update_session_knowledge_bases(session, id, kb_ids, user_id=current_user.id)
    if not updated:
        raise HTTPException(status_code=404, detail="Session not found")
    return updated

@router.get("/sessions/{id}")
def get_session_details(
    id: int,
    session: Session = Depends(get_session),
    current_user: User = Depends(get_current_user)
) -> Any:
    result = code_chat_service.get_session(session, id, user_id=current_user.id)
    if not result:
        raise HTTPException(status_code=404, detail="Session not found")
    return result
